### ISSUE TITLE
Testing aborted programmatically. Time limit was reached. (architectures: arm64, ppc64el, riscv64, s390x)

### ISSUE THREAD

**Issue by elboulangero** [AFLplusplus/AFLplusplus#2056]
Hello,

I'm seeing a timeout in `test-llvm.sh`, logs are at https://buildd.debian.org/status/fetch.php?pkg=aflplusplus&arch=ppc64el&ver=4.20c-1&stamp=1713174746&raw=0

The relevant parts:

```
    Test case count : 1 favored, 0 variable, 0 ignored, 1 total
       Bitmap range : 5 to 5 bits (average: 5.00 bits)
        Exec timing : 0 to 0 us (average: 0 us)

[*] No -t option specified, so I'll use an exec timeout of 20 ms.
[+] All set and ready to roll!
[*] Entering queue cycle 1.

[...]

+++ Testing aborted programmatically +++
Time limit was reached
Statistics: 6 new corpus items found, 50.00% coverage achieved, 0 crashes saved, 0 timeouts saved, total runtime 0 days, 0 hrs, 0 min, 30 sec
We're done here. Have a nice day!
CUT------------------------------------------------------------------CUT
[!] afl-fuzz is not working correctly with llvm_mode cmplog
```

It took me a while to locate the exact test that fails, but finally I found it:

```
$ git grep 'with llvm_mode cmplog'
test/test-llvm.sh:275:      $ECHO "$GREEN[+] afl-fuzz is working correctly with llvm_mode cmplog"
test/test-llvm.sh:280:      $ECHO "$RED[!] afl-fuzz is not working correctly with llvm_mode cmplog"
```

Looking at the git history, initially this test was with `-V60` and it was changed to `-V15` and then to `-V30`: https://github.com/AFLplusplus/AFLplusplus/commit/9bc5abc4eca13996fccc05b4ad08b7ac5a217e25#diff-4ab711ded7c01d5a4ea10ced307cfbd4f7a420f88d93dcc14c6c4cb226fc2712L266

- - - - 

Back to the failure on the Debian infrastructure: the test failed on architecture `ppc64el`, it could be that it was running on a slow or overworked machine.

Am I right that I should revert from `-V30` to `-V60` to prevent this test from being flaky on slow machines? That's my understanding at this point, but TBH it's more guessing than understanding ;) So I'd better ask.

Thanks!

**Comment by vanhauser-thc** [comment]
That is just a switch how long the fuzzing should be done. You can try -V60 if that works. But it is also possible that there is an llvm ppc bug that makes this test fail. LLVM has quite some bugs.

**Comment by elboulangero** [comment]
I did run a few builds on a `ppc64el` box. Unit tests failed once, but for a different reason (I didn't keep the logs).

Looking at the build history for this package, for the `ppc64el` architecture, at https://buildd.debian.org/status/logs.php?pkg=aflplusplus&arch=ppc64el: It failed two times. Ok it's not zero, but not worth investigating either.

So I'm closing the issue, feel free to re-open if you wish.

**Comment by elboulangero** [comment]
I'm re-opening the ticket, the problem surfaced again during the last rebuild of the Debian package, however this time 4 architectures failed to build, cf https://buildd.debian.org/status/package.php?p=aflplusplus, screenshot below:

![Screenshot from 2024-05-13 09-47-37](https://github.com/AFLplusplus/AFLplusplus/assets/1942580/93b24574-be35-4752-9da5-75b91061dc1d)

The build failure on `s390` is something else, out of scope here. But for the 3 other architectures, it's the same build failure:

```
    Test case count : 1 favored, 0 variable, 0 ignored, 1 total
       Bitmap range : 5 to 5 bits (average: 5.00 bits)
        Exec timing : 0 to 0 us (average: 0 us)

[*] No -t option specified, so I'll use an exec timeout of 20 ms.
[+] All set and ready to roll!
[*] Entering queue cycle 1.

[...]

+++ Testing aborted programmatically +++
Time limit was reached
Statistics: 6 new corpus items found, 50.00% coverage achieved, 0 crashes saved, 0 timeouts saved, total runtime 0 days, 0 hrs, 0 min, 30 sec
We're done here. Have a nice day!
CUT------------------------------------------------------------------CUT
[!] afl-fuzz is not working correctly with llvm_mode cmplog
```

Build logs:
- arm64: https://buildd.debian.org/status/fetch.php?pkg=aflplusplus&arch=arm64&ver=4.20c-3&stamp=1715417851&raw=0
- ppc64el: https://buildd.debian.org/status/fetch.php?pkg=aflplusplus&arch=ppc64el&ver=4.20c-3&stamp=1715417456&raw=0
- riscv64: https://buildd.debian.org/status/fetch.php?pkg=aflplusplus&arch=riscv64&ver=4.20c-3&stamp=1715418586&raw=0

**Comment by elboulangero** [comment]
I reverted the duration of the fuzzing test from 30 seconds to 60 seconds (as discussed above). Let's see if that improves anything. If you have any better suggestion, happy to hear!

**Comment by vanhauser-thc** [comment]
there is no platform specific code in the cmplog module. if at least one platform succeeds then a failure means that either there is a bug in llvm for those platforms or the build machines (because of emulation and load) is running slow. if it is the latter than increasing the test time will help.

**Comment by vanhauser-thc** [comment]
hmm but looking at the output it looks fishy. it should have enough time solving it but does not solve it.
is it possible that you send me this compiled `test-cmplog` binary, best from arm64:

`AFL_LLVM_CMPLOG=1 afl-clang-fast -o test-cmplog test/test-cmplog.c`

**Comment by elboulangero** [comment]
Another round of builds in the Debian infrastructure, and I still get 3 "Time limit was reached", but on different architectures, this time it's `armel`, `i386` and `ppc64el`.

![Screenshot from 2024-05-13 14-58-35](https://github.com/AFLplusplus/AFLplusplus/assets/1942580/da8383c2-a2aa-4d56-abdf-597392f7ef1f)

> is it possible that you send me this compiled test-cmplog binary, best from arm64:
>
> AFL_LLVM_CMPLOG=1 afl-clang-fast -o test-cmplog test/test-cmplog.c

Will do

**Comment by elboulangero** [comment]
> is it possible that you send me this compiled test-cmplog binary, best from arm64:
>
> AFL_LLVM_CMPLOG=1 afl-clang-fast -o test-cmplog test/test-cmplog.c

Here it is, attached and zipped.
[test-cmplog.zip](https://github.com/AFLplusplus/AFLplusplus/files/15291840/test-cmplog.zip)

**Comment by vanhauser-thc** [comment]
![Untitled](https://github.com/AFLplusplus/AFLplusplus/assets/7396244/67210960-e0c7-4d60-9fe6-3902c6122783)
yeah the problem is that cmplog cannot resolve the comparison. it is instrumented correctly so this is not an llvm bug.
No idea what goes wrong.
I am installing an arm64 qemu debian and will check it out but this will take some time

**Comment by vanhauser-thc** [comment]
I installed a Debian arm64 qemu environment and found that the solving time needed is way too high for a slow emulated system. I removed those parts of the test case that have nothing to do with cmplog and now it completes for me in 12 second.
the change is now  in the dev branch

**Comment by elboulangero** [comment]
Thanks! I suppose it's https://github.com/AFLplusplus/AFLplusplus/commit/a87ea969133d9d403fd5207ec08a1ffdc4d8376b? I'm trying it now

**Comment by vanhauser-thc** [comment]
Yes

**Comment by elboulangero** [comment]
Hello again! It looks much better now at https://buildd.debian.org/status/package.php?p=aflplusplus, screenshot below:

![Screenshot from 2024-05-15 09-02-23](https://github.com/AFLplusplus/AFLplusplus/assets/1942580/f65058f2-8e5c-4e4d-a986-8fa002c9041a)

The riscv64 arch still failed with the same timeout though: https://buildd.debian.org/status/fetch.php?pkg=aflplusplus&arch=riscv64&ver=4.20c-5&stamp=1715710745&raw=0. Do you want that I send you another test-cmplog for this architecture, as I did in https://github.com/AFLplusplus/AFLplusplus/issues/2056#issuecomment-2107001458? 

Or should I increase the test timeout for this architecture (and maybe also s390x, it didn't build yet, probably because the builder is overworked)?

**Comment by vanhauser-thc** [comment]
I installed various Debian qemu instances.

riscv works for me:
```
...
[+] llvm_mode instrumentlist feature works correctly
[*] running afl-fuzz for llvm_mode cmplog, this will take approx 10 seconds
[+] afl-fuzz is working correctly with llvm_mode cmplog
[+] llvm_mode persistent mode feature works correctly
[*] 1 test cases completed.
[+] all test cases executed
[+] all tests were successful :-)
root@debian:/afl/test# uname -a
Linux debian 6.7.12-riscv64 #1 SMP Debian 6.7.12-1 (2024-04-24) riscv64 GNU/Linux
```
**but** LTO mode does not work because the binutils package does not contain ld.gold and llvm 18 from apt.llvm.org does not have /usr/lib/llvm-18/lib/LLVMgold.so

s390:
```
...
[*] llvm_mode laf-intel/compcov testing splitting integer types (this might take
 some time)
[ 4014.555557] User process fault: interruption code 003b ilc:2 in libc.so.6[3ff
84300000+1d5000]
[ 4014.557729] Failing address: 000003fb8447c000 TEID: 000003fb8447c800
[ 4014.557779] Fault in primary space mode while using user ASCE.
[ 4014.557836] AS:00000000027081c7 R3:0000000000000024 
[ 4014.557982] CPU: 0 PID: 10638 Comm: test-compcov.co Not tainted 6.7.12-s390x 
#1  Debian 6.7.12-1
```
this is a bug in llvm :)
cmplog mode works with another tweak I just pushed.

**Comment by elboulangero** [comment]
## riscv64

> riscv works for me [...] but LTO mode does not work because the binutils package does not contain ld.gold and llvm 18 from apt.llvm.org does not have /usr/lib/llvm-18/lib/LLVMgold.so

I suppose you refer to this part of the [build logs](https://buildd.debian.org/status/fetch.php?pkg=aflplusplus&arch=riscv64&ver=4.20c-5&stamp=1715710745&raw=0) that is repeated in various places:

```
[+] llvm_mode detected llvm 12+, enabling afl-lto LTO implementation
GNUmakefile.llvm:241: clang option -flto is not working - maybe LLVMgold.so not found - cannot enable LTO mode
```

There's a bug opened: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1068225

## s390

> s390 [...] this is a bug in llvm

The [build logs](https://buildd.debian.org/status/fetch.php?pkg=aflplusplus&arch=s390x&ver=4.20c-5&stamp=1715822693&raw=0) have a different output, but maybe it's just because you were running with some verbose flag enabled. In the build logs from the Debian builder there is only:

```
[*] llvm_mode laf-intel/compcov testing splitting integer types (this might take some time)
[!] llvm_mode laf-intel/compcov feature compilation failed
```

>  cmplog mode works with another tweak I just pushed

You mean this one: https://github.com/AFLplusplus/AFLplusplus/commit/0a16ea74879a935bbeb33e488cd451d6d9a7e19a ?

**Comment by vanhauser-thc** [comment]
> ## riscv64
> > riscv works for me [...] but LTO mode does not work because the binutils package does not contain ld.gold and llvm 18 from apt.llvm.org does not have /usr/lib/llvm-18/lib/LLVMgold.so
> 
> I suppose you refer to this part of the [build logs](https://buildd.debian.org/status/fetch.php?pkg=aflplusplus&arch=riscv64&ver=4.20c-5&stamp=1715710745&raw=0) that is repeated in various places:
> 
> ```
> [+] llvm_mode detected llvm 12+, enabling afl-lto LTO implementation
> GNUmakefile.llvm:241: clang option -flto is not working - maybe LLVMgold.so not found - cannot enable LTO mode
> ```
> 
> There's a bug opened: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1068225

I do not refer to the build logs but what I saw on my local qemu instance :) but yes.
but the bug report is only 50% of the problem.
The other is that the binutils package does not contain ld.gold (at least for the debian I have installed)

> ## s390
> > s390 [...] this is a bug in llvm
> 
> The [build logs](https://buildd.debian.org/status/fetch.php?pkg=aflplusplus&arch=s390x&ver=4.20c-5&stamp=1715822693&raw=0) have a different output, but maybe it's just because you were running with some verbose flag enabled. In the build logs from the Debian builder there is only:
> 
> ```
> [*] llvm_mode laf-intel/compcov testing splitting integer types (this might take some time)
> [!] llvm_mode laf-intel/compcov feature compilation failed
> ```

the error I pasted is only visible in dmesg (or on the console if your system config prints kernel errors there).

> > cmplog mode works with another tweak I just pushed
> 
> You mean this one: [0a16ea7](https://github.com/AFLplusplus/AFLplusplus/commit/0a16ea74879a935bbeb33e488cd451d6d9a7e19a) ?

yes

**Comment by elboulangero** [comment]
>  The other is that the binutils package does not contain ld.gold (at least for the debian I have installed)

A quick search, and people seem to say that ld.gold for risc64 simply doesn't exist, and is not under development:
- https://gitlab.alpinelinux.org/alpine/aports/-/issues/15904
- https://discussion.fedoraproject.org/t/fedora-riscv64-binutils-2-42-instead-of-2-41-for-the-rest/104634/3
- https://discussion.fedoraproject.org/t/fedora-riscv-should-we-disable-lto-gcc-already-disabled-for-clang/96582

**Comment by vanhauser-thc** [comment]
Ah ok so there will be no af-clang-lto for riscv then :)

**Comment by elboulangero** [comment]
Hello!

So I applied [0a16ea7](https://github.com/AFLplusplus/AFLplusplus/commit/0a16ea74879a935bbeb33e488cd451d6d9a7e19a) and uploaded a new package last week. The build for `s390x` [still fails](https://buildd.debian.org/status/fetch.php?pkg=aflplusplus&arch=s390x&ver=4.20c-6&stamp=1715874091&raw=0):

```
[*] llvm_mode laf-intel/compcov testing splitting integer types (this might take some time)
[!] llvm_mode laf-intel/compcov feature compilation failed
```

Now that I read again the conversation, I wonder: the patch 0a16ea7 was supposed to fix that, or was it supposed to fix something else? You said « _ cmplog mode works with another tweak I just pushed_ », and I don't know if it's related to the laf-intel/compcov failure above.

I tried a rebuild on a s390x box and I can reproduce the issue. Unfortunately I don't have access to kernel logs on this box, so I can't confirm if it's the same error as the one you mentioned in https://github.com/AFLplusplus/AFLplusplus/issues/2056#issuecomment-2111947967. I'm trying to setup a s390x QEMU VM, but no success so far.

One last detail: looking at the logs history at https://buildd.debian.org/status/logs.php?pkg=aflplusplus&arch=s390x, we can see it started to fail from 2024-05-11, and comparing logs, the biggest change in the environment is the lib that went from 2.37 to 2.38. So it seems that the issue is caused by libc 2.38.

**Comment by vanhauser-thc** [comment]
The issue in s390x is a bug in llvm and cannot be fixed by us. (I wrote as much above).

I have a s390x qemu box set up:
```
wget -O qemu.zip https://gitlab.com/api/v4/projects/giomasce%2Fdqib/jobs/artifacts/master/download?job=convert_s390x-virt
unzip qemu.zip
cd dqib_s390x-virt
qemu-system-s390x -machine 's390-ccw-virtio' -cpu max -m 1G -drive file=image.qcow2 -device virtio-net-ccw,netdev=net -netdev user,id=net,hostfwd=tcp::2223-:22 -kernel kernel -initrd initrd -nographic -append "root=LABEL=rootfs console=ttyAMA0"
```
accounts:
```
root:root
debian:debian
```

**Comment by elboulangero** [comment]
Thanks for all the time you spent on that!

I opened https://github.com/AFLplusplus/AFLplusplus/issues/2097 for clarity, to document the issue discovered with s390x.

And I'm closing this one since the test timeout issues were fixed.

Feel free to close https://github.com/AFLplusplus/AFLplusplus/issues/2097 as well if you want, as you said it's a bug in llvm and not in aflplusplus.

Thanks again!
